File: leetcode/113. Path Sum II.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        res = []
        def traverse(node, curr_sum, curr_lst):
            if node is None:
                return
            curr_sum += node.val
            curr_lst.append(node.val)
            if node.left is None and node.right is None:
                if curr_sum == targetSum:
                    res.append(curr_lst[:])
            traverse(node.left, curr_sum, curr_lst)
            traverse(node.right, curr_sum, curr_lst)
            curr_lst.pop()
        
        traverse(root, 0, [])
        return res

File: leetcode/113. Path Sum II.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun pathSum(root: TreeNode?, targetSum: Int): List<List<Int>> {
        val res = mutableListOf<MutableList<Int>>()
        if (root == null) return res

        fun traverse(node: TreeNode?, remain: Int, path: MutableList<Int>) {
            if (node == null) return
            val remain: Int = remain - node.`val`
            path.add(node.`val`)
            if (remain == 0 && node.left == null && node.right == null) {
                res.add(path.toMutableList())
            }
            traverse(node.left, remain, path)
            traverse(node.right, remain, path)
            path.removeLast()
        }
        traverse(root, targetSum, mutableListOf<Int>())

        return res
    }
}
